crates/sui-data-ingestion-core/src/reducer.rs

```diff
@@ -0,0 +1,50 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{Reducer, Worker};
+use anyhow::Result;
+use std::collections::HashMap;
+use sui_types::messages_checkpoint::CheckpointSequenceNumber;
+use tokio::sync::mpsc;
+
+pub(crate) async fn reduce<W: Worker>(
+    task_name: String,
+    mut current_checkpoint_number: CheckpointSequenceNumber,
+    mut progress_receiver: mpsc::Receiver<(CheckpointSequenceNumber, W::Result)>,
+    executor_progress_sender: mpsc::Sender<(String, CheckpointSequenceNumber)>,
+    reducer: Option<Box<dyn Reducer<W::Result>>>,
+) -> Result<()> {
+    let mut unprocessed = HashMap::new();
+    let mut batch = vec![];
+    let mut progress_update = None;
+
+    while let Some((checkpoint_number, message)) = progress_receiver.recv().await {
+        unprocessed.insert(checkpoint_number, message);
+        while let Some(message) = unprocessed.remove(&current_checkpoint_number) {
```

Comment:
following the idea of `should_commit`, I think we can quickly get all available elements from the stream, and then call should_commit, where
1. for fastest commit like indexer primary worker, it will return the whole batch and we will commit all of them
2. for both snapshot lag / epoch split, `should_commit` should return a slice, the batch will remove the slice and leave the rest for next iteration
3. for archives that requires some "completeness check", here we will keep the whole batch until the condiiton is met

wdyt?

Reply:
yeah, my first version of implementation was similar to that with a partial commit and ability to return unhandled items back to the reducer. Brandon pointed out that it might be confusing and error proning for developers